hack/make.sh

```diff
@@ -118,12 +118,16 @@ if \
 fi
 
 # test whether "libdevmapper.h" is new enough to support deferred remove
-# functionality.
+# functionality. We favour libdm_dlsym_deferred_remove over
+# libdm_no_deferred_remove in dynamic cases because the binary could be shipped
+# with a newer libdevmapper than the one it was built wih. hack/make/daemon
+# will change to libdm_no_deferred_remove as dlsym doesn't make sense for
+# static binaries.
 if \
 	command -v gcc &> /dev/null \
 	&& ! ( echo -e  '#include <libdevmapper.h>\nint main() { dm_task_deferred_remove(NULL); }'| gcc -xc - -o /dev/null $(pkg-config --libs devmapper) &> /dev/null ) \
 ; then
-	DOCKER_BUILDTAGS+=' libdm_no_deferred_remove'
+	DOCKER_BUILDTAGS+=' libdm_dlsym_deferred_remove'
 fi
```

Comment:
Perhaps it makes sense to only do the check and add the tag only in case `DOCKER_BUILDTAGS` does not already contain anything that looks like `libdm_*_deferred_remove` (or so). In other words, try to detect whether we need `libdm_dlsym_deferred_remove` only if `libdm_WHATEVER_deferred_remove` is not yet set.

What do you think?

Reply:
Like this?

```
tags_contains_libdm_deferred_remove=""
for tag in $DOCKER_BUILDTAGS
do
	if [[ "$tag" == libdm_*_deferred_remove ]]
	then
		tags_contains_libdm_deferred_remove=1
	fi
done
if [ -z "$tags_contains_libdm_deferred_remove" ]
then
	# test whether "libdevmapper.h" is new enough to support deferred remove
	# functionality. We favour libdm_dlsym_deferred_remove over
	# libdm_no_deferred_remove in dynamic cases because the binary could be shipped
	# with a newer libdevmapper than the one it was built wih.
	if \
		command -v gcc &> /dev/null \
		&& ! ( echo -e  '#include <libdevmapper.h>\nint main() { dm_task_deferred_remove(NULL); }'| gcc -xc - -o /dev/null $(pkg-config --libs devmapper) &> /dev/null ) \
	; then
		DOCKER_BUILDTAGS+=' libdm_dlsym_deferred_remove'
	fi
fi
```

Seems like a bit ... overkill (to me at least)? But it's not *that* horrible, I guess.

Reply:
Yes, the idea here is if someone specifies what to do about libdm explicitly we won't override it.

Can be done in a cleaner fashion I think...

```shell
# Adds <prefix>_<tag> to DOCKER_BUILDTAGS unless it already
# contains a word starting from <prefix>_
add_buildtag() {
	local prefix=$1
	local tag=$2

	case " $DOCKER_BUILDTAGS" in
		*" ${prefix}_"*)
			;;
		*)
			DOCKER_BUILDTAGS="$DOCKER_BUILDTAGS ${prefix}_${tag}"
			;;
	esac
}
....

add_buildtag libdm dlsym_deferred_remove
```

the above code is POSIX sh compatible. Since we're using bash anyway, it can be boiled down to:

```bash
# Adds $1_$2 to DOCKER_BUILDTAGS unless it already
# contains a word starting from $1_
add_buildtag() {
        [[ " $DOCKER_BUILDTAGS" == *" $1_"* ]] || DOCKER_BUILDTAGS+=" $1_$2"
}
```